Sources/System/FilePath/FilePathSyntax.swift

```diff
@@ -0,0 +1,353 @@
+  /// Returns `nil` if the path is empty or only contains a root
+  ///
+  /// REVIEW NOTE: We follow  `basename` semantics, unlike `file_name` in
+  /// C++ which treats trailing `/` specially, or in Rust which returns nil for `..`
+  ///
+  /// REVIEW NOTE: shell and C basename will return `/` if the path is `/`. I don't
+  /// think we want to, it seems clearer that this only is in the relative portion of the path,
+  /// and doing so would make the setter weird.
+  public var basename: FilePath.Component? {
+    get {
+      let (rootEnd, _) = _parseRoot()
+      guard let base = self.components.last, base.slice.startIndex >= rootEnd else {
+        return nil
+      }
+      return base
+    }
+    set {
+      // TODO: this is assuming that pop correctly preserves root
+      _ = self.pop()
+      if let c = newValue { self.append(FilePath(c)) }
+    }
+  }
+
+  /// Returns the path up to but not including the `basename`.
+  ///
+  /// If the path only contains a root, returns `self`.
+  /// If the path has no root and only includes a single component,
+  /// returns an empty FilePath.
+  ///
+  /// REVIEW NOTE: Shell returns `.` if there is no dirname, while we return the empty path.
+  /// REVIEW NOTE: I removed the setter. It seems sketchy and its use indicates programmer error.
+  public var dirname: FilePath {
+    get {
+      // No basename means path is root or empty (which is the dirname)
+      if self.basename == nil { return self }
+      return FilePath(self.components.dropLast())
+    }
+  }
+
+  /// The extension of the file or directoy last component.
+  ///
+  /// If `basename` is `nil` or one of the special path components
+  /// `.` or `..`, `get` returns `nil` and `set` does nothing.
+  ///
+  /// If `basename` does not contain a `.` anywhere, or only
+  /// at the start, `get` returns `nil` and `set` will append a
+  /// `.` and `newValue` to `basename`
+  ///
+  /// Otherwise `get` returns everyting after it, and `set` will
+  /// replace the extension
+  ///
+  /// Examples:
+  ///   * `/tmp/foo.txt => "txt"`
+  ///   * `/Appliations/Foo.app/ => app`
+  ///   * `/Appliations/Foo.app/bar.txt => txt`
+  ///   * `/tmp/.hidden => nil``
+  ///   * `/tmp/.. => nil``
+  ///
+  /// REVIEW NOTE: Rust has a set_extension instead, which returns whether it
+  /// did anything, but having a setter is more convenient.
+  public var `extension`: String? {
+    get {
+      fatalError()
+    }
+    set { fatalError() }
+  }
+
+  /// The non-extension portion of the file or directoy last component.
+  ///
+  /// Returns `nil` if `basename` is `nil`
+  ///
+  ///   * `/tmp/foo.txt => "foo"`
+  ///   * `/Appliations/Foo.app/ => Foo`
+  ///   * `/Appliations/Foo.app/bar.txt => bar`
+  ///   * `/tmp/.hidden => .hidden``
+  ///   * `/tmp/.. => ..``
+  ///   * `/ => nil`
+  ///
+  /// TODO: setter? That's a little harder to imagine than for `extension`.
+  /// Setting to `nil` is more likely a bug.
+  public var stem: String? {
```

Comment:
Since it's only the root that doesn't have a stem, it feels a little odd to have this be optional, because of the effect on the callsites.